File: Edu/tree/ValidBinarySearchTree.js
import { TreeNode, treeToArray } from "./TreeNode.js";

const root = new TreeNode(
    5,
    new TreeNode(4),
    new TreeNode(6,
        new TreeNode(3),
        new TreeNode(7)
    )
);

const root2 = new TreeNode(
    3,
    new TreeNode(1,
        null,
        new TreeNode(2)

    ),
    new TreeNode(4)
);

class Solution {
    /**
     * @param {TreeNode} root
     * @return {boolean}
     */
    isValidBST(root) {
        if (!root) true;

        return this.isValidBSTRec(root.left, -Number.MAX_VALUE, root.val)
            && this.isValidBSTRec(root.right, root.val, Number.MAX_VALUE);
    }

    isValidBSTRec(root, min, max) {
        if (!root) return true;

        return root.val > min && root.val < max && this.isValidBSTRec(root.left, min, root.val)
            && this.isValidBSTRec(root.right, root.val, max);
    }
}

console.log(treeToArray(
    root
));

console.log(
    new Solution().isValidBST(root)
);

console.log(
    new Solution().isValidBST(root2)
);

File: java/src/com/leetcode/tree/ValidBinarySearchTree.java
package com.leetcode.tree;

import static com.leetcode.tree.TreeNode.arrayToTree;

public class ValidBinarySearchTree {

    public static void main(String[] args) {
        System.out.println(
                new ValidBinarySearchTree().isValidBST(
                    arrayToTree(new Integer[] {2,1,3})
                )
        );
    }

    public boolean isValidBST(TreeNode root) {
        return isValidBST(root, Long.MIN_VALUE, Long.MAX_VALUE);
    }

    private boolean isValidBST(TreeNode root, long min, long max) {
        if (root == null) return true;

        return root.val > min && root.val < max
                && isValidBST(root.left, min, root.val)
                && isValidBST(root.right, root.val, max);
    }
 }
